backend/common/repo.py

```diff
@@ -0,0 +1,204 @@
+"""
+Repository tools
+"""
+# coding: utf-8
+import typing
+import http
+import os
+import zlib
+import lzma
+from urllib import parse
+import hashlib
+
+import requests
+
+# pylint:disable=W0612,W0212,C0301
+
+
+class GeneralRepoException(Exception):
+    """
+    Dpkg repository exception
+    """
+
+
+class DpkgRepo:
+    """
+    Dpkg repository detection.
+    The repositories in Debian world have several layouts,
+    such as "flat", classic tree, PPA etc.
+    """
+    PKG_GZ = "Packages.gz"
+    PKG_XZ = "Packages.xz"
+    PKG_RW = "Packages"
+
+    class ReleaseEntry:  # pylint: disable=W0612,R0903
+        """
+        Release file entry
+        """
+        class Checksum:  # pylint: disable=R0903
+            """
+            Checksums of the Release file
+            """
+            md5: str = ""
+            sha1: str = ""
+            sha256: str = ""
+
+        def __init__(self, size: int, uri: str):
+            self.checksum = DpkgRepo.ReleaseEntry.Checksum()
+            self.size = size
+            self.uri = uri
+
+    class EntryDict(dict):
+        """
+        Parsed release container.
+        """
+        def __init__(self, repo: "DpkgRepo"):
+            super().__init__()
+            self.__repo = repo
+
+        def get(self, key: typing.Any) -> typing.Optional[typing.Any]:
+            if not self.__repo.is_flat():
+                key = "/".join(parse.urlparse(self.__repo._url).path.strip("/").split("/")[-2:] + [key])
+            return self[key]
+
+    def __init__(self, url):
+        self._url = url
+        self._flat = None
+        self._pkg_index = None, None
+        self._release = DpkgRepo.EntryDict(self)
+
+    def get_pkg_index_raw(self) -> typing.Tuple[str, bytes]:
+        """
+        Get Packages.gz or Packages.xz or Packages content, raw.
+
+        :return: bytes of the content
+        """
+        if self._pkg_index[0] is None:
+            for cnt_fname in [DpkgRepo.PKG_GZ, DpkgRepo.PKG_XZ, DpkgRepo.PKG_RW]:
+                packages_url = os.path.join(self._url, cnt_fname)
```

Comment:
Just found a problem. SCC repo urls are something like `https://updates.suse.com/SUSE/Updates/Ubuntu/16.04-CLIENT-TOOLS/x86_64/update/?very_long_auth_token`.
The auth token should be split and added at the end, i.e. `https://updates.suse.com/SUSE/Updates/Ubuntu/16.04-CLIENT-TOOLS/x86_64/update/Packages.gz?very_long_auth_token`.

Reply:
Ah, you are right. I will also have to add some unit tests here as well.